### PR TITLE
Numbers - unit, zero, eps, numtostr

### PR DESCRIPTION
**Pull request by renehiemstra**
For all number types: primitives, nfloats, complex numbers:
* `T:unit()`, `T:zero()`
* `tmath.numtostr(x : T)`

For all primitive types:
* `T:eps()`

`T:unit()`, `T:zero()`, `T:eps()` are lua methods that return a terra constant variable.

### CODE REVIEW

File: nfloat.t [hunk 1]
```diff
@@ -101,13 +101,34 @@ local terra shiftandscale(n : uint64, e : int)
...
 end
 
 
 local FixedFloat = terralib.memoize(function(N)
+
+    --float_type[N] stores the high-precision number using the following layout
+    --local M = N / 64
+    --struct float_type
+    --  head : uint64[2]
+    --  d : uint64[M-1]
```
Comment (tpkessler) [inline nfloat.t:114]: That should be `uint64[M]`?

File: nfloat.t [hunk 2]
```diff
@@ -122,6 +143,39 @@ local FixedFloat = terralib.memoize(function(N)
...
+        for i = 1, M do
+            d[i] = 0ULL
+        end
+        if value == 0 then
+            return {{9223372036854775808ULL, 0ULL}, d}
+        elseif value == 1 then
+            d[M] = 9223372036854775808ULL
+            return {{1ULL, 0ULL}, d}
+        elseif value == "eps" then
+            d[M] = 9223372036854775808ULL
```
Comment (tpkessler) [inline nfloat.t:162]: Is this [uint64(-1)?]

Comment (renehiemstra) [reply nfloat.t:162]: Right

File: nfloat.t (hunk 1 above)
Comment (renehiemstra) [reply nfloat.t:114]: Yes

File: nfloat.t (hunk 2 above)
Comment (renehiemstra) [reply nfloat.t:162]: No

Comment (tpkessler) [reply nfloat.t:162]: Can you add an explanation for this?

File: complex.t [hunk 3]
```diff
@@ -62,24 +70,45 @@ local complex = terralib.memoize(function(T)
...
+    terra complex:tostr()
+        var str : int8[16]
+        var re, im =  self:real(), self:imag()
+        if im < 0 then
+            im = -im
+            var s1, s2 = tmath.numtostr(re), tmath.numtostr(im)
+            C.strcpy(&str[0], s1)
+            C.strcat(&str[0], "-")
+            C.strcat(&str[0], s2)
+            C.strcat(&str[0], "im")
```
[the remark this answers is not in the document]
Comment (renehiemstra) [reply complex.t]: replaced with `snprintf`

File: nfloat.t (hunk 2 above)
Comment (renehiemstra) [reply nfloat.t]: replaced it with a library call to `bit.lshift`

File: complex.t [hunk 4]
```diff
@@ -89,26 +92,27 @@ local complex = terralib.memoize(function(T)
         tmath.abs:adddefinition(terra(x: complex) return x:norm() end)
     end
 
+    --maxlen is twice the size of T and twice one char for the sign
```
Comment (tpkessler) [inline complex.t:95]: Don't forget the closing `\0` that ends C strings.

File: mathfuns.t [hunk 5]
```diff
@@ -94,94 +94,103 @@ for tname, cname in pairs(funs_three_var) do
...
-for _, T in ipairs{int8, int16, int32, int64} do
-    local impl = terra(v : T)
-        var str : int8[8]
-        C.sprintf(&str[0], "%d", v)
-        return str
-    end
-    math.numtostr:adddefinition(impl)
+--determine the mximum number of digits in representing a number of
+--n bytes. here we use:
+--log10(2) \approx 0.31
```
Comment (tpkessler) [inline mathfuns.t:162]: Outdated docs

File: nfloat.t [hunk 6]
```diff
@@ -153,13 +157,14 @@ local FixedFloat = terralib.memoize(function(N)
         for i = 1, M do
             d[i] = 0ULL
         end
+        local bitshiftone = bit.lshift(1ULL, 63)
```
Comment (tpkessler) [inline nfloat.t:160]: Can you please add an explanation for the `63`?

File: nfloat.t [hunk 7]
```diff
@@ -312,9 +317,18 @@ local FixedFloat = terralib.memoize(function(N)
...
-        return tmath.numtostr(v:truncatetodouble())
-    end)
+    --for now we format up to double precision.
+    --ToDo: specialized print.
+    local format = global(rawstring, "%0.2f")
+    local maxlen = tmath.ndigits(sizeof(double))
+    tmath.numtostr:adddefinition(
+        terra(v : nfloat)
+            var buffer : int8[maxlen]
+            var j = C.snprintf(buffer, maxlen, format, v:truncatetodouble())
```
Comment (tpkessler) [inline nfloat.t:327]: Unused return value

File: template.t [hunk 8]
```diff
@@ -314,15 +314,15 @@ function Template:new()
 
 	function template:adddefinition(methods)
 		methods = methods or {}
-		for sig,func in pairs(methods) do
+		for sig, func in pairs(methods) do
 			--check if method with this serialized key already exists
 			for s,v in pairs(self.methods) do
 				if sig:serialize()==s:serialize() then --overwrite old definition
-					self.methods[s] = func
+					self.methods[s] = terralib.memoize(func)
```
Comment (tpkessler) [inline template.t:321]: Better to define `func = terralib.memoize(func)` at the top of the loop body.

File: test_complex.t [hunk 9]
```diff
@@ -12,6 +12,26 @@ local nfloat = require("nfloat")
 
 local float256 = nfloat.FixedFloat(256)
 
+if not __silent__ then
```
Comment (tpkessler) [inline test_complex.t:15]: Why not comparing the strings in a unit test? You can use `strcmp()` from `string.h`.

File: test_math.t [hunk 10]
```diff
@@ -6,6 +6,23 @@
 import "terratest/terratest"
 
 local tmath = require('mathfuns')
+local io = terralib.includec("stdio.h")
+
+if not __silent__ then
```
Comment (tpkessler) [inline test_math.t:11]: See my comment on `test_complex.t`

File: test_nfloat.t [hunk 11]
```diff
@@ -5,9 +5,28 @@
 
 import "terratest/terratest"
 
+local io = terralib.includec("stdio.h")
 local nfloat = require("nfloat")
 local tmath = require("mathfuns")
 
+if not __silent__ then
```
Comment (tpkessler) [inline test_nfloat.t:12]: See my comment in `test_complex.t`.

File: test_complex.t (hunk 9 above)
Comment (renehiemstra) [reply test_complex.t:15]: ok

File: test_nfloat.t (hunk 11 above)
Comment (renehiemstra) [reply test_nfloat.t:12]: ok

File: template.t (hunk 8 above)
Comment (renehiemstra) [reply template.t:321]: good point.

File: nfloat.t (hunk 6 above)
Comment (renehiemstra) [reply nfloat.t:160]: explained

File: mathfuns.t (hunk 5 above)
Comment (renehiemstra) [reply mathfuns.t:162]: ok

File: complex.t (hunk 4 above)
Comment (renehiemstra) [reply complex.t:95]: `snprintf` takes care of this.

**Pull request merged**